File: SOEN3014 工程项目实践（卓越工程师型）/课程作业/大数据检索速度竞赛/代码（Spark）/src/main/scala/SparkSolution.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.functions.*
import org.apache.spark.sql.types.{LongType, StructField, StructType}

import java.nio.file.{Files, Path}

object SparkSolution extends App :

  val DATASET_SCALE = DatasetScale.BIG
  val OUTPUT_PATH = "data/result.txt"

  val start = System.currentTimeMillis()

  val spark = SparkSession
    .builder()
    .appName("SparkSolution")
    .master("local[*]")
    .getOrCreate()

  val paths = List(DATASET_SCALE.DATA_FOLDER_A, DATASET_SCALE.DATA_FOLDER_B)
    .flatMap(folder => Range(1000, 1000 + DATASET_SCALE.FILE_NUM).map(i => s"$folder/part_$i"))

  val schema = StructType(List(
    StructField("c1", LongType, nullable = false),
    StructField("c2", LongType, nullable = false),
    StructField("c3", LongType, nullable = false)
  ))
  val df = spark.read.schema(schema).option("delimiter", "|").csv(paths *)

  val window = Window.partitionBy("sum")
  val result = df
    .withColumn("sum", col("c1") + col("c2") + col("c3"))
    .withColumn("dupeCount", count("*").over(window))
    .where(col("dupeCount") > 1)
    .drop("dupeCount")
    .sort(col("sum"))
    .collect()

  var end = System.currentTimeMillis()
  println(s"Spark spent: ${(end - start) / 1000.0} s")

  Files.write(
    Path.of(OUTPUT_PATH),
    result
      .map(row => s"${row(3)} ${row(0)}|${row(1)}|${row(2)}")
      .mkString("\n")
      .getBytes
  )

  end = System.currentTimeMillis()
  println(s"Total time spent: ${(end - start) / 1000.0} s")

File: SOEN3014 工程项目实践（卓越工程师型）/课程作业/大数据检索速度竞赛/代码（Spark）/src/main/scala/spark_solution.py
"""Pyspark solution for the problem.
"""

import time
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, count
from pyspark.sql.types import StructType, StructField, LongType
from pyspark.sql.window import Window

from dataset_scale import DatasetScale

if __name__ == '__main__':

    DATASET_SCALE = DatasetScale.BIG
    OUTPUT_PATH = "data/result.txt"

    start = time.time()

    spark = (SparkSession
             .builder
             .appName("PysparkSolution")
             .master("local[*]")
             .getOrCreate())

    paths = []
    for folder in (DATASET_SCALE.value.data_folder_a,
                   DATASET_SCALE.value.data_folder_b):
        for i in range(1000, 1000 + DATASET_SCALE.value.file_num):
            paths.append(f'{folder}/part_{i}')

    schema = StructType([
        StructField("c1", LongType(), nullable=False),
        StructField("c2", LongType(), nullable=False),
        StructField("c3", LongType(), nullable=False)
    ])
    df = spark.read.schema(schema).option("delimiter", "|").csv(paths)

    window = Window.partitionBy("sum")
    result = (df
              .withColumn("sum", col("c1") + col("c2") + col("c3"))
              .withColumn("dupeCount", count("*").over(window))
              .where(col("dupeCount") > 1)
              .drop("dupeCount")
              .sort(col("sum"))
              .collect())

    end = time.time()
    print(f'Spark spent: {(end - start) / 1000} s')
    print(f'{time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())}')

    with open(OUTPUT_PATH, 'w', encoding='utf-8') as f:
        f.writelines(f'{row[3]} {row[0]}|{row[1]}|{row[2]}\n' for row in result)
